### ISSUE TITLE
h5py lock-up on array load

### ISSUE THREAD

**Issue by martindurant** [dask/distributed#860]
When rewriting the dask-tutorial to use distributed, find that the dask-array example with a hdf5 file freezes, presumably in lock contention. I have not isolated the problem, but it does not happen with the default scheduler. I unravelled da.from_array a bit to try to make it clearer

This first version passes, where `dset` is global
```python
fn = 'random.hdf5' # generated by dask-tutorial/prep.py, random_array
f = h5py.File(fn, mode='r')
dset = f['/x']

def get_part(off):
    return dset[off:off+1000000]

@gen_cluster(client=True)
def test_hdf(c, s, a, b):
    parts = [delayed(get_part)(r) for r in range(0, 10000000, 1000000)]
    bits = [da.from_delayed(part, shape=(1000000, ), dtype='f4')
            for part in parts]
    x = da.concatenate(bits)
    f = c.compute(x.sum())
    out = yield c._gather(f)
```

but if you pass `dset`, it fails (dset itself pickles OK)
```python
def get_part(off, dset):
    return dset[off:off+1000000]

@gen_cluster(client=True)
def test_hdf_direct(c, s, a, b):
    f = h5py.File(fn, mode='r')
    dset = f['/x']

    parts = [delayed(get_part)(r, dset) for r in range(0, 10000000, 1000000)]
    bits = [da.from_delayed(part, shape=(1000000, ), dtype='f4')
            for part in parts]
    x = da.concatenate(bits)
    f = c.compute(x.sum())
    out = yield c._gather(f)
```

**Comment by seibert** [comment]
Not sure whether you are running multiple threads in the worker process, but the HDF5 library has some global state that isn't thread-safe (even when accessing files read-only) unless it has been built with `--enable-threadsafe`.  Last I checked, the Anaconda build of HDF5 doesn't turn on this flag as it also can reduce performance (IIRC).

**Comment by mrocklin** [comment]
from_array should introduce a lock by default

**Comment by martindurant** [comment]
Even calling from_array with lock=True still showed the error.

**Comment by martindurant** [comment]
Interestingly, I am not seeing the problem with `dataframe.read_hdf`

**Comment by mrocklin** [comment]
Looking at this example more closely now.  So the problem here is that h5py objects aren't picklable (nor are they ever likely to be).  Fortunately, we have defined our own serialization for these objects, but that only works if the objects are top-level in the graph

    {'x': dset}  # this should be ok
    {'x': (myfunc, dset)}  # this would fail

So a quick fix to your test would probably be to call `delayed` on `dset` before passing it into another function.

Your original problem with `da.from_array` should handle this just fine.  The `from_array` function puts the original array into the graph at the top level. 

We actually test this in `distributed/protocol/tests/test_h5py.py`:

```python
@gen_cluster(client=True)
def test_h5py_serialize(c, s, a, b):
    from dask.utils import SerializableLock
    lock = SerializableLock('hdf5')
    with tmpfile() as fn:
        with h5py.File(fn, mode='a') as f:
            x = f.create_dataset('/group/x', shape=(4,), dtype='i4',
                                 chunks=(2,))
            x[:] = [1, 2, 3, 4]
        with h5py.File(fn, mode='r') as f:
            dset = f['/group/x']
            x = da.from_array(dset, chunks=dset.chunks, lock=lock)
            y = c.compute(x)
            y = yield y._result()
            assert (y[:] == dset[:]).all()
```

However it seemed like you ran into a problem with `da.from_array` in the wild.  Are you able to retrigger that failure in an isolated test?  I'm curious what is different between your situation and the passing test copied above.

**Comment by martindurant** [comment]
It looks very similar to the test above
```python
@gen_cluster(client=True)
def test_hdf(c, s, a, b):
    f = h5py.File(fn, mode='r')
    dset = f['/x']
    x = da.from_array(dset, chunks=(1000000,), lock=True)
    f = c.compute(x.sum())
    out = yield c._gather(f)
```
It fails when providing a pre-made SerializableLock too. `dset.chunks` is None.

**Comment by mrocklin** [comment]
I can get the other tests to fail if there is only a single chunk.  Task fusion seems to be the culprit.

**Comment by mrocklin** [comment]
So a less-than-ideal solution for tutorial materials may be to use a smaller chunksize such that there are multiple chunks.  (if this is actually the problem)

**Comment by martindurant** [comment]
There are multiple chunks: the total data-size is (1000000000,). 
In your test, `dset` automatically has a value for chunks, but not here.

**Comment by mrocklin** [comment]
This might reproduce the problem: https://github.com/dask/distributed/pull/861

**Comment by mrocklin** [comment]
I believe that this has been resolved by #861
